**Retrive metrics of specific branches in ubiquitous genes OGs**

---

In this tutorial we will compare through a Likelihood Ratio Test (LRT from now on)
a model with a single omega values across all the branch of our phylogeny versus a model in which all the branches have different omegas, for each OG; subsequently 
we will extract omega (along with other metrics) for specific branches. This analysis will focus only on [OGs of ubiquitus genes](https://github.com/for-giobbe/BASE/tree/master/example/_ubiquitous_OGs),
in which a single-copy gene is present for each species.
After moving to the [toy-dataset folder](https://github.com/for-giobbe/BASE/tree/master/example) we can quickly revise what's needed to start our analysis:

* a [folder](https://github.com/for-giobbe/BASE/tree/master/example/_ubiquitous_OGs) containing aligned OGs, in fasta format and with the ```.fa``` extenction. 
These alignment shouldn't have any STOP codon, yet BASE will report and exclude the OGs wehre they are found. The fasta alignment headers have to be identical to the tips names in the phylogenetic tree.

* a [species tree](https://github.com/for-giobbe/BASE/blob/master/example/sp.tre) - in the newick format ```.nwk```. It has to include all the species considered and
no branchlengths are needed as they will be optimized for each gene through our analysis. 
The tips names in the phylogenetic tree have to be identical to the headers of the fasta alignements.

* two codeml ```.ctl``` files, which describe the models we want to leverage in our analysis.

In the latter ```.ctl``` files, every parameter can still be modified, but the ```seqfile =```, ```outfile =``` , ```treefile =``` fields should be left empty.
As stated before, in this analysis we will compare two branch models:
a [model](https://github.com/for-giobbe/BASE/blob/master/example/m0.ctl) where there is one omega shared by all branches and (codeml model 0) and
a [model](https://github.com/for-giobbe/BASE/blob/master/example/m1.ctl) where each branch has it's own omega (codeml model 1). 
When using your own data, remember to properly set the genetic code in the ```.ctl``` files. 

Moreover remember that species name should match exactly between OGs and phylogeny.

---

We can start the analysis using 4 cores - using ```--cores``` 4 - and restricting the analysis to ubiquitous OGs - using the ```--ubiquitous``` flag:

```
	sh ../BASE.sh --analyze --input _ubiquitous_OGs/ --output _ubiquitous_OGs_0VS1 
	--tree spp_tree.nwk --model_a m0.ctl --model_b m1.ctl 
	--cores 4 --ubiquitous
```

Some information are printed to the standard output, including potential errors, as can be seen from the second-last line:

```
  analysis started on Fri Sep 25 14:02:08 CEST 2020

  analyizing 1 replicate(s) of 21 genes: branch models 0 VS 1 & site models 0 VS 0 

  analyze:	100/100 

  performing LRT 

  formatting output 

  the analysis has produced some warnings: you will find the information relative to each failure in the file warning_summary.txt

  analysis finished on Fri Sep 25 14:07:19 CEST 2020 
```

Inside the  output folder ```_ubiquitous_OGs_0VS1``` you will find a ```warnings_summary.txt```.
If you ```cat``` the file, you can read all the warnings generated by the analysis, such as:
```the OG OG3105.fa is non-ubiquitous and has been excluded by the analyses```
This is because this OG misses one of the 11 species which are present in the analysis species tree; we'll see how to included such OGs furhter in the tutorial.
The analysis generated a codeml output (```.out```) for each OG: theses files names state whether the general or alternative model was
the best-fit and - if specified - the best replicate (more on that in following tutorials). We'll need these outputs for the following step, but a
rather important output is allready the summary of the LRTs. Type ```column -t column -t _ubiquitous_OGs_0VS1/likelihood_summary.txt``` to take a look at it:

```
OG      branch_model_A  site_model_A  model_A_np  model_A_LnL    rep_A  branch_model_B  site_model_B  model_B_np  model_B_LnL   rep_B  LRT      df  p.value  significance
OG3126  0               0             2           -4132.017209   1      1               0             12          -4116.793123  1      30.4482  10  7e-04    ***
OG3158  0               0             2           -3818.030135   1      1               0             12          -3801.882145  1      32.296   10  4e-04    ***
OG3164  0               0             2           -5374.980383   1      1               0             12          -5358.727983  1      32.5048  10  3e-04    ***
OG3196  0               0             2           -3365.98055    1      1               0             12          -3355.771198  1      20.4187  10  0.0255   *
OG3197  0               0             2           -3680.783168   1      1               0             12          -3671.481767  1      18.6028  10  0.0456   *
OG3302  0               0             2           -9054.676043   1      1               0             12          -9028.37395   1      52.6042  10  0        ***
OG3342  0               0             2           -3444.803353   1      1               0             12          -3431.378852  1      26.849   10  0.0028   **
OG3347  0               0             2           -9597.616313   1      1               0             12          -9586.571185  1      22.0903  10  0.0147   *
OG3359  0               0             2           -3147.226951   1      1               0             12          -3126.908131  1      40.6376  10  0        ***
OG3362  0               0             2           -5722.861684   1      1               0             12          -5706.727938  1      32.2675  10  4e-04    ***
OG3372  0               0             2           -3710.675722   1      1               0             12          -3695.63184   1      30.0878  10  8e-04    ***
OG3387  0               0             2           -4067.125785   1      1               0             12          -4057.24774   1      19.7561  10  0.0316   *
OG3395  0               0             2           -3901.33739    1      1               0             12          -3841.403883  1      119.867  10  0        ***
OG3399  0               0             2           -3003.011552   1      1               0             12          -2990.356715  1      25.3097  10  0.0048   **
OG3600  0               0             2           -2867.120583   1      1               0             12          -2844.358272  1      45.5246  10  0        ***
OG3622  0               0             2           -3407.557493   1      1               0             12          -3388.589881  1      37.9352  10  0        ***
OG3640  0               0             2           -3342.883393   1      1               0             12          -3317.158544  1      51.4497  10  0        ***
OG3648  0               0             2           -10491.352418  1      1               0             12          -10484.2421   1      14.2206  10  0.1632   n/s
OG3682  0               0             2           -3703.868866   1      1               0             12          -3690.236972  1      27.2638  10  0.0024   **
OG3683  0               0             2           -2791.481942   1      1               0             12          -2774.797116  1      33.3697  10  2e-04    ***
```

This output is self-explanatory and the more important information in this table are the result of the LRTs found in the last two columns.

---

Subsequently we can proceed to the further step: ```--extract```. it
takes as input the folder generated by the previous step - which containt the codeml outputs - and will generate
an equivalent number of  ```.annotation ``` files, which contain for each OG:

* a tree with internal nodes matching codeml outputs
* the species which are associated to each branch

Type:

```
    sh ../BASE.sh --extract --input _ubiquitous_OGs_0VS1
```

and here goes the standard output:

```
  analysis started on Fri Sep 25 14:58:43 CEST 2020

  annotating 20 codeml outputs 

  annotate:	100% 

  annotation ok 

  label file is missing - relaunch the analysis with label(s) to extract specific branches/clades

  analysis finished on Fri Sep 25 15:02:59 CEST 2020 
```

As the secon-last line tells us, perfroming this step without specifying any branch/clade will just  annotate codeml outputs;
to retrive the metrics relative to our branch/clade of interest we need to specify them using a file like [this](https://github.com/for-giobbe/BASE/blob/master/example/branch.lst) 
which contains on each line all the species associated to our branch/clade of interested - separated by single spaces - followed by a custom identifier.
Here ```branch_of_interest``` and ```species_of_interest``` are used but they can be changed to any name. Let's take a look to our labels file:

```
dmag dpul branch_of_interest
dmag species_of_interest
``` 

We also need to specify a minimum number of species to consider a branch/clade: the latter can be expressed with an
absolute number - such as 3 - or via a proportion - such as 0.9 - but when we want to retrive metrics only when all the species of our branch/clade of interest
are presen we can just specify ```--min_spp x```. 

Let's extract the metrics relative to our branch of interest, using the line:

```
    sh ../BASE.sh --extract --input _ubiquitous_OGs_0VS1 --labels branch.lst --min_spp x
```

This will print to the screen:

```
  analysis started on Fri Sep 25 15:12:50 CEST 2020

  annotation ok 

  extracting 2 branches from 20 codeml output 

  extract branch_o..	 100%
  extract single_s..	 100% 
 
  the analysis has produced some warning: you will find the information relative to each failure in the file warning_summary.txt 
 
  analysis finished on Fri Sep 25 15:14:11 CEST 2020 
```

The warning printed to the standard output is the same generated in the analyze step, so we shouldn't care about it.
The two separate ```--extract``` commands could have been carried out also in a single step, by specifying since the beginning the labels file. 
If one wants to extract additional branches - either modifying prehexisting label file of using a novel one - this step will be faster,
as the workflow will recognize the ```.annotation``` files and skip that step.
We can then see the output by typing```column -t _ubiquitous_OGs_0VS1/branch.clade_of_interest.min.spp.1.dNdS.summary```; as you can see
each line of the label file generates a summary output, named with the identifyer and the treshold of missing species 
(min.spp meaning minimum of species). This is an ouput for a internal branch:

```
branch/clade        gene    spp_n  dNdS    t      dN      dS
branch_of_interest  OG3126  2      0.0646  0.518  0.0403  0.6234
branch_of_interest  OG3158  2      0.0946  0.482  0.0508  0.5369
branch_of_interest  OG3164  2      0.0866  0.425  0.0474  0.5469
branch_of_interest  OG3196  2      0.1364  0.505  0.0651  0.4774
branch_of_interest  OG3197  2      0.1076  0.639  0.0688  0.6394
branch_of_interest  OG3302  2      0.0974  0.392  0.0438  0.4493
branch_of_interest  OG3342  2      0.0837  0.389  0.0380  0.4536
branch_of_interest  OG3347  2      0.0697  0.375  0.0323  0.4628
branch_of_interest  OG3359  2      0.0272  0.324  0.0109  0.4007
branch_of_interest  OG3362  2      0.1145  0.398  0.0465  0.4057
branch_of_interest  OG3372  2      0.0231  0.318  0.0107  0.4639
branch_of_interest  OG3387  2      0.1054  0.292  0.0295  0.2797
branch_of_interest  OG3395  2      0.0655  0.218  0.0170  0.2595
branch_of_interest  OG3399  2      0.0919  0.625  0.0640  0.6967
branch_of_interest  OG3600  2      0.1416  0.207  0.0293  0.2067
branch_of_interest  OG3622  2      0.0684  0.530  0.0432  0.6312
branch_of_interest  OG3640  2      0.1045  0.411  0.0464  0.4440
branch_of_interest  OG3648  2      0.1708  0.471  0.0726  0.4250
branch_of_interest  OG3682  2      0.1242  0.277  0.0348  0.2803
branch_of_interest  OG3683  2      0.0993  0.407  0.0448  0.4517
```

Typing ```column -t _ubiquitous_OGs_annotate/branch.species_of_interest.dNdS.summary``` 
you can see how an output for a terminal branch looks like:

```
species         gene    dNdS    t      dN      dS
single_species  OG3126  0.0646  0.085  0.0066  0.1023
single_species  OG3158  0.0946  0.109  0.0115  0.1215
single_species  OG3164  0.0866  0.155  0.0173  0.1999
single_species  OG3196  0.1364  0.269  0.0346  0.2539
single_species  OG3197  0.1076  0.286  0.0308  0.2864
single_species  OG3302  0.0974  0.144  0.0161  0.1649
single_species  OG3342  0.0837  0.118  0.0115  0.1370
single_species  OG3347  0.0697  0.223  0.0192  0.2754
single_species  OG3359  0.0272  0.020  0.0007  0.0251
single_species  OG3362  0.1145  0.141  0.0165  0.1439
single_species  OG3372  0.0231  0.095  0.0032  0.1392
single_species  OG3387  0.1054  0.102  0.0103  0.0981
single_species  OG3395  0.0655  0.104  0.0081  0.1231
single_species  OG3399  0.0919  0.148  0.0151  0.1647
single_species  OG3600  0.1416  0.064  0.0091  0.0644
single_species  OG3622  0.0684  0.187  0.0153  0.2232
single_species  OG3640  0.1045  0.159  0.0180  0.1718
single_species  OG3648  0.1708  0.125  0.0193  0.1127
single_species  OG3682  0.1242  0.094  0.0118  0.0950
single_species  OG3683  0.0993  0.090  0.0099  0.0997
```

These tables are rather self explanatory as well, and represent the final output of our analysis.

---

[Back](https://github.com/for-giobbe/BASE/blob/master/tutorial_0.md) to the tutorials list.
