**Retrive omega values of specific branches in ubiquitous genes OGs**

---

In this tutorial we will compare through a Likelihood Ratio Test (LRT from now on)
a model with a single omega values across all the branch of our phylogeny versus a model in which all the branches have different omegas, for each OG; subsequently 
we will extract omega (along with dN dS and substitution rate) for specific branches. This analysis will focus only on [OGs of ubiquitus genes](https://github.com/for-giobbe/BASE/tree/master/example/_ubiquitous_OGs),
in which a single-copy gene is present for each species.
After moving to the [toy-dataset folder](https://github.com/for-giobbe/BASE/tree/master/example) we can quickly revise what's needed to start our analysis:

* a folder containing aligned OGs ```.aln```. These alignment shouldn't have any STOP codon, 
yet BASE will report the OGs wehre they are found.

* a species tree - in the newick format ```.nwk```. It has to include all the species considered;
no branchlengths are needed as they will be optimized for each gene through our analysis.

* two codeml ```.ctl``` files, which describe the models we want to leverage in our analysis. 

In the latter ```.ctl``` files, every parameter can still be modified, but the ```seqfile =```, ```outfile =``` , ```treefile =``` fields should	be left	empty.
As stated before, in this analysis we will compare two branch models:
a [model](https://github.com/for-giobbe/BASE/blob/master/example/m0.ctl) where there is one omega shared by all branches and (codeml model 0) and
a (model)[https://github.com/for-giobbe/BASE/blob/master/example/m1.ctl] where each branch has it's own omega (codeml model 1). 
When using your own data, remember to properly set the genetic code in the ```.ctl``` files.

---

To start the analysis just use the line:

```
    BASE.sh --analyze -i _ubiquitous_OGs/ -o _ubiquitous_OGs_analyze 
    -t sp.tre -ma m0.ctl -mb m1.ctl -c 4
```

Some information are printed to the standard output, including potential errors, as can be seen from the second-last line:

```
  analysis started on Fri Sep 25 14:02:08 CEST 2020

  analyizing 1 replicate(s) of 21 genes: model 0 VS 1 & NSsites 0 VS 0 

  analyze:	100/100 

  performing LRT 

  formatting output 

  the analysis has produced some warning: you will find the information relative to each failure in the file warning_summary.txt

  analysis finished on Fri Sep 25 14:07:19 CEST 2020 
```

Inside the  output folder ```_ubiquitous_OGs_analyze``` you will find a ```warnings_summary.txt```.
If you ```cat``` the file, you can read all the warnings generated by the analysis, such as:
```the gene OG3105.mafft.n.aln contains missing data and has been excluded by the analyses - to include it use the -d flag```
This is because an OG misses one of the 11 species which are present in the species tree. This OG can be included as we'll se in the further tutorial.
The analysis generated a codeml output (```.out```) for each OG. Theses files names state whether the general or alternative model was
the best-fit and - if specified - the best replicate. We'll need these outputs for the following steps.
A rather important file is the summary of the LRTs. Type ```column -t _ubiquitous_OGs_analyze/likelihood_summary.txt``` to take a look at it:

```
Ortholog_Cluster  Model1  NSsites_a  Model1np  Model1LnL      Rep1  Model2  NSsites_a.1  Model2np  Model2LnL     Rep2  LRT      df  p.value  significance
OG3126            0       NA         2         -4132.017209   1     1       NA           12        -4116.792933  1     30.4486  10  7e-04    ***
OG3158            0       NA         2         -3818.030135   1     1       NA           12        -3801.882145  1     32.296   10  4e-04    ***
OG3164            0       NA         2         -5374.980383   1     1       NA           12        -5358.727983  1     32.5048  10  3e-04    ***
OG3196            0       NA         2         -3365.98055    1     1       NA           12        -3355.771214  1     20.4187  10  0.0255   *
OG3197            0       NA         2         -3680.783168   1     1       NA           12        -3671.481767  1     18.6028  10  0.0456   *
OG3302            0       NA         2         -9054.676043   1     1       NA           12        -9028.37395   1     52.6042  10  0        ***
OG3342            0       NA         2         -3444.803353   1     1       NA           12        -3431.378852  1     26.849   10  0.0028   **
OG3347            0       NA         2         -9597.616313   1     1       NA           12        -9586.571185  1     22.0903  10  0.0147   *
OG3359            0       NA         2         -3147.226951   1     1       NA           12        -3126.908131  1     40.6376  10  0        ***
OG3362            0       NA         2         -5722.861684   1     1       NA           12        -5706.727938  1     32.2675  10  4e-04    ***
OG3372            0       NA         2         -3710.675722   1     1       NA           12        -3695.63184   1     30.0878  10  8e-04    ***
OG3387            0       NA         2         -4067.125785   1     1       NA           12        -4057.24774   1     19.7561  10  0.0316   *
OG3395            0       NA         2         -3901.33739    1     1       NA           12        -3841.403883  1     119.867  10  0        ***
OG3399            0       NA         2         -3003.011552   1     1       NA           12        -2990.356715  1     25.3097  10  0.0048   **
OG3600            0       NA         2         -2867.120583   1     1       NA           12        -2844.358272  1     45.5246  10  0        ***
OG3622            0       NA         2         -3407.557493   1     1       NA           12        -3388.589881  1     37.9352  10  0        ***
OG3640            0       NA         2         -3342.883393   1     1       NA           12        -3317.158544  1     51.4497  10  0        ***
OG3648            0       NA         2         -10491.352418  1     1       NA           12        -10484.2421   1     14.2206  10  0.1632   n/s
OG3682            0       NA         2         -3703.868866   1     1       NA           12        -3690.236972  1     27.2638  10  0.0024   **
OG3683            0       NA         2         -2791.481942   1     1       NA           12        -2774.797116  1     33.3697  10  2e-04    ***
```

This output is quite self-explanatory, and the more important outcome of this table is the result of the LRTs which are found in the last two columns.

---

Subsequently we can proceed to the further step, ```annotate```, whcih
takes as input the folder generated by the previous step and will write a new folder - specified through the ```-o``` flag
In the output folder you will find the original  ```.out ``` and also some  ```.result ```, which contains

* the internal nodes annotation created by codeml
* the tips which are associated to each phylogeny branch. 

Type:

```
    sh BASE.sh --annotate -i _ubiquitous_OGs_analyze/ -o _ubiquitous_OGs_annotate
```

and here goes the standard output:

```
  analysis started on Fri Sep 25 14:58:43 CEST 2020

  annotating 20 codeml outputs 

  annotate:	100/100
  
  analysis finished on Fri Sep 25 15:02:59 CEST 2020 
```

---

In the last step we proceed to extract the values relative to the branch(es) which we are interested in. 
We will specify each internal branch using a file like [this](https://github.com/for-giobbe/BASE/blob/master/example/branch.lst) which contains on each line all the species which
form the clade relative to our branch of interested - separated by single spaces - followed by an identifier.
If we are interested in a terminal branch relative to a single tip, just specify the tip name in the tree and an identifier.
let's take a look to such file:

```
dmag dpul clade_of_interest
dmag species_of_interest
``` 

We also need to specify the minimum number of species to consider a clade: the latter can be expressed with an
absolute number - such as 3 - or via a proportion - such as 0.9. When we are not considering missing data we can just specify ```-n 1```. 
Let's extract the values of the branch leading to the two species dmag and dpul, using the line:

```
    sh BASE.sh --extract -i _ubiquitous_OGs_annotate -l branch.lst -n 1
```

This will print to the screen:

```
  analysis started on Fri Sep 25 15:12:50 CEST 2020

  extracting 2 branches from 20 codeml output 
  
  extract clade_o..	 10/100

  extract species..	 10/100
 
  analysis finished on Fri Sep 25 15:14:11 CEST 2020 
```

We can then see the output by typing```column -t _ubiquitous_OGs_annotate/branch.clade_of_interest.min.otu.1.dNdS.summary```; as you can see
each line of the label file generates an output, named with the name of their clade /tip and the treshold of missing data 
(min.otu, where otu stands for Operational Taxonomic Units). This is an ouput for a clade:

```
clade              gene    OTUs_n  dNdS    t      dN      dS
clade_of_interest  OG3126  2       0.0978  0.518  0.0559  0.5711
clade_of_interest  OG3158  2       0.1264  0.482  0.0628  0.4966
clade_of_interest  OG3164  2       0.1314  0.425  0.0630  0.4796
clade_of_interest  OG3196  2       0.1297  0.505  0.0629  0.4844
clade_of_interest  OG3197  2       0.1381  0.639  0.0826  0.5982
clade_of_interest  OG3302  2       0.1491  0.392  0.0588  0.3945
clade_of_interest  OG3342  2       0.1238  0.389  0.0507  0.4097
clade_of_interest  OG3347  2       0.0735  0.375  0.0336  0.4580
clade_of_interest  OG3359  2       0.0299  0.324  0.0119  0.3978
clade_of_interest  OG3362  2       0.1541  0.398  0.0573  0.3717
clade_of_interest  OG3372  2       0.0219  0.318  0.0102  0.4656
clade_of_interest  OG3387  2       0.0953  0.292  0.0272  0.2857
clade_of_interest  OG3395  2       0.0476  0.218  0.0130  0.2727
clade_of_interest  OG3399  2       0.1391  0.625  0.0864  0.6212
clade_of_interest  OG3600  2       0.2059  0.207  0.0370  0.1798
clade_of_interest  OG3622  2       0.0965  0.530  0.0566  0.5867
clade_of_interest  OG3640  2       0.1355  0.411  0.0559  0.4127
clade_of_interest  OG3648  2       0.1708  0.471  0.0726  0.4250
clade_of_interest  OG3682  2       0.1398  0.277  0.0378  0.2705
clade_of_interest  OG3683  2       0.1825  0.407  0.0679  0.3720
```

Typing ```column -t _ubiquitous_OGs_annotate/branch.single_species.min.otu.1.dNdS.summary``` you can see how an output for a terminal branche should look like:

```
clade           gene    dNdS    t      dN      dS
single_species  OG3126  0.0201  0.085  0.0023  0.1171
single_species  OG3158  0.0603  0.109  0.0081  0.1336
single_species  OG3164  0.0424  0.155  0.0098  0.2321
single_species  OG3196  0.0571  0.269  0.0174  0.3055
single_species  OG3197  0.0705  0.286  0.0220  0.3122
single_species  OG3302  0.0476  0.144  0.0091  0.1906
single_species  OG3342  0.0844  0.118  0.0116  0.1371
single_species  OG3347  0.0618  0.223  0.0174  0.2820
single_species  OG3359  0.0001  0.020  0.0000  0.0272
single_species  OG3362  0.0981  0.141  0.0147  0.1497
single_species  OG3372  0.0036  0.095  0.0005  0.1492
single_species  OG3387  0.0749  0.102  0.0078  0.1047
single_species  OG3395  0.0001  0.104  0.0000  0.1499
single_species  OG3399  0.0530  0.148  0.0097  0.1832
single_species  OG3600  0.1323  0.064  0.0087  0.0658
single_species  OG3622  0.0198  0.187  0.0051  0.2584
single_species  OG3640  0.0613  0.159  0.0118  0.1930
single_species  OG3648  0.1708  0.125  0.0193  0.1127
single_species  OG3682  0.0294  0.094  0.0036  0.1219
single_species  OG3683  0.0290  0.090  0.0035  0.1220
```

These tables are rather self explanatory as well, and represent the final output of our analysis.

---

[Back](https://github.com/for-giobbe/BASE/blob/master/tutorial_0.md) to the tutorials list.
